experimental/iterators/include/utils/tuple.h

```diff
@@ -0,0 +1,103 @@
+#ifndef UTILS_PRINT_H
+#define UTILS_PRINT_H
+
+#include <iostream>
+#include <tuple>
+
+namespace impl {
+template <typename TupleType, std::size_t... kIndices>
+auto extractElements(const TupleType &tuple,
+                     const std::index_sequence<kIndices...> & /*unused*/) {
+  if constexpr (sizeof...(kIndices) > 0) {
+    return std::make_tuple(std::get<kIndices>(tuple)...);
+  } else {
+    return std::make_tuple();
+  }
+}
+
+template <std::size_t kHeadSize, typename TupleType, std::size_t... kIndices>
+auto extractTail(const TupleType &tuple,
+                 const std::index_sequence<kIndices...> & /*unused*/) {
+  using IndexSequence =
+      std::integer_sequence<std::size_t, kIndices + kHeadSize...>;
+  return extractElements(tuple, IndexSequence{});
+}
+} // namespace impl
+
+/// Extracts the first `kHeadSize` elements from the given tuple.
+template <std::size_t kHeadSize, typename... Types>
+auto extractHead(const std::tuple<Types...> &tuple) {
+  using TupleType = std::tuple<Types...>;
+  using IndexSequence = std::make_index_sequence<kHeadSize>;
+  return impl::extractElements(tuple, IndexSequence{});
+}
+
+/// Extracts the last `kTailSize` elements from the given tuple.
+template <std::size_t kHeadSize, typename... Types>
+auto extractTail(const std::tuple<Types...> &tuple) {
+  using TupleType = std::tuple<Types...>;
+  static constexpr std::size_t kTailSize = sizeof...(Types) - kHeadSize;
+  using IndexSequence = std::make_index_sequence<kTailSize>;
+  return impl::extractTail<kHeadSize>(tuple, IndexSequence{});
+}
+
+namespace impl {
+template <std::size_t kIndex, typename TupleType>
+std::size_t hashTupleField(const TupleType &tuple) {
+  std::hash<std::tuple_element_t<kIndex, TupleType>> hasher;
+  return hasher(std::get<kIndex>(tuple));
+}
+
+template <typename TupleType, std::size_t... kIndices>
+std::size_t hashTuple(const TupleType &tuple,
+                      const std::index_sequence<kIndices...> & /*unused*/) {
+  return (hashTupleField<kIndices>(tuple) ^ ...);
```

Comment:
fancy template-fu :)

Reply:
Yes, but I think it is necessary (if we want statically typed and inlinable code). It's amazingly powerful once you got your head around it though ;)